### PR TITLE
[diem-operational-tool] Add x25519 Key generation and peer info

### PR DESCRIPTION
**Pull request by gregnazario**
### Overview 
A tool to generate x25519 keys, and extract information necessary to be
used for allowlisting on other nodes as trusted peers.

#### Generating private keys, and building peer from it's public key
```
$ cargo run -p diem-operational-tool -- generate-key --key-file /tmp/greg-is-great.txt --encoding hex --key-type x25519
{
  "Result": "Success"
}
$ cargo run -p diem-operational-tool -- extract-peer-from-file --key-file /tmp/greg-is-great.txt --encoding hex --output-file /tmp/greg-is-awesome.txt
{
  "Result": {
    "367002dae32e50c3583ab4c3bc4bc923": {
      "addresses": [],
      "keys": [
        "de02328535cf05d3fa7cac6dcacf833a367002dae32e50c3583ab4c3bc4bc923"
      ],
      "role": "Downstream"
    }
  }
}
$ cat /tmp/greg-is-awesome.txt 
---
367002dae32e50c3583ab4c3bc4bc923:
  addresses: []
  keys:
    - de02328535cf05d3fa7cac6dcacf833a367002dae32e50c3583ab4c3bc4bc923
  role: Downstream

```

#### Building peers from list of public keys
```
$ cargo run -p diem-operational-tool -- extract-peers-from-keys --keys 0000000000000000111111111111111122222222222222223333333333333333,1111111111111111222222222222222233333333333333334444444444444444 --output-file /tmp/testing-is-great.txt
{
  "Result": {
    "22222222222222223333333333333333": {
      "addresses": [],
      "keys": [
        "0000000000000000111111111111111122222222222222223333333333333333"
      ],
      "role": "Downstream"
    },
    "33333333333333334444444444444444": {
      "addresses": [],
      "keys": [
        "1111111111111111222222222222222233333333333333334444444444444444"
      ],
      "role": "Downstream"
    }
  }
}

$ cat /tmp/testing-is-great.txt 
---
"22222222222222223333333333333333":
  addresses: []
  keys:
    - "0000000000000000111111111111111122222222222222223333333333333333"
  role: Downstream
"33333333333333334444444444444444":
  addresses: []
  keys:
    - "1111111111111111222222222222222233333333333333334444444444444444"
  role: Downstream

```

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
Few comments:
* Maybe make this general to generate a key specified by the type parameter and default to Ed25519
* We should update the ExtractPublicKey and ExtractPrivateKey to support that as well, since some folks might use Vault for handling their key
* I don't think we want to extract them necessarily as bcs, since the config expects them to be in hex, no? or should we have another function for converting between hex, bcs, and maybe base64
* Also I imagine the generate config would actually take in a list of keys, but I guess one could argue either way

These are some thoughts, I don't think there's a perfect solution here but it feels worth exploring the design space.

**Comment by gregnazario** [comment]
> * Maybe make this general to generate a key specified by the type parameter and default to Ed25519

This makes sense, I'm not sure we should default to either value, as they will output the same causing confusion.

> * We should update the ExtractPublicKey and ExtractPrivateKey to support that as well, since some folks might use Vault for handling their key

Yeah, let's get this all looped together in common code

> * I don't think we want to extract them necessarily as bcs, since the config expects them to be in hex, no? or should we have another function for converting between hex, bcs, and maybe base64

Let's serialize these into hex.  I believe bcs was just chosen because we use it, but I don't see it being directly used.  Base64 would be useful for how some parts are encoded in helm, but I'll check with the PEs.

> * Also I imagine the generate config would actually take in a list of keys, but I guess one could argue either way

Yeah, this isn't really generating a config, just converting them.  Directly generating a config gets into the question of what kind of peers am I generating this peer for (PeerRole).  I'll think about how best to approach this.

**Comment by gregnazario** [comment]
Note: I'm working on combining the copied code on key generation and all, but it's taking a bunch of time, and now doesn't build correctly so I'm skipping that right now.

File: config/management/operational/src/keys.rs [hunk 1]
```diff
@@ -1,15 +1,159 @@
...
+
+impl FromStr for KeyType {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "Ed25519" | "ed25519" => Ok(KeyType::Ed25519),
+            "X25519" | "x25519" => Ok(KeyType::X25519),
+            _ => Err("Invalid key type"),
+        }
```
Comment (davidiw) [inline config/management/operational/src/keys.rs:52]: why not be extra fancy and just convert it all to lower case and do that match?

File: config/management/operational/src/keys.rs [hunk 2]
```diff
@@ -1,15 +1,159 @@
...
+    Hex,
+}
+
+impl FromStr for EncodingType {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "hex" | "Hex" => Ok(EncodingType::Hex),
+            "bcs" | "Bcs" | "BCS" => Ok(EncodingType::BCS),
```
Comment (davidiw) [inline config/management/operational/src/keys.rs:68]: same... to lower saves the day

Comment (davidiw) [reply config/management/operational/src/keys.rs:68]: could we add base64, because of reasons?

File: config/management/operational/src/keys.rs [hunk 3]
```diff
@@ -20,6 +164,8 @@ struct ExtractKey {
     /// Location to store the key
     #[structopt(long)]
     key_file: PathBuf,
+    #[structopt(long)]
+    encoding: EncodingType,
```
Comment (davidiw) [inline config/management/operational/src/keys.rs:168]: hmm... this is a breaking change, can you do a check on the partners repo to see if this function is used anywhere?

**Review by davidiw — COMMENTED** [review]
Overall looks good! Very good.

File: config/management/operational/src/keys.rs (hunk 1 above)
Comment (gregnazario) [reply config/management/operational/src/keys.rs:52]: I legitimately forgot about it :D

File: config/management/operational/src/keys.rs (hunk 3 above)
Comment (gregnazario) [reply config/management/operational/src/keys.rs]: Eh, I just am making it default to BCS.  It'll be it's main use case anyways

File: config/management/operational/src/keys.rs (hunk 2 above)
Comment (gregnazario) [reply config/management/operational/src/keys.rs]: Sure

**Comment by mgorven** [comment]
* What does this look like in the fullnode configuration file? Ideally we'd output in the exact format needed so operators can just copy/paste.
* Can the same peer ID and key be used in multiple fullnode instances?

**Comment by gregnazario** [comment]
> * What does this look like in the fullnode configuration file? Ideally we'd output in the exact format needed so operators can just copy/paste.

This should be the exact same format (in YAML) as the Network config file.  It just might need to be tabbed inward to fit in the `seeds` field.

> * Can the same peer ID and key be used in multiple fullnode instances?

Yes, and no.  If you reuse the same PeerId between multiple nodes, and they connect to the same node, one of them will lose.  PeerId is the unique identifier of a connection, and it would need to be different per node.

By default, PeerId must be derived by public key to access on a public network, so you could just create a new key through these tools per node.  This doesn't scale very well for now, and is part of the improvements to be made in the future.

**Comment by gregnazario** [comment]
> > * What does this look like in the fullnode configuration file? Ideally we'd output in the exact format needed so operators can just copy/paste.

Woops, I have an in progress change to combine multiple keys into one big YAML, it's just not yet done

**Comment by gregnazario** [comment]
> * What does this look like in the fullnode configuration file? Ideally we'd output in the exact format needed so operators can just copy/paste.

Updated above to show what it comes out as with the full YAML.

Remember, that to collect them all at once, you'd need all the public keys.  Individual ones are not as exciting, because each node may have different storage.

**Comment by davidiw** [comment]
> $ cat /tmp/greg-is-awesome.txt 

Perhaps.

File: config/management/operational/src/keys.rs [hunk 4]
```diff
@@ -1,15 +1,221 @@
...
+            _ => Err("Invalid key type"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, StructOpt)]
+pub enum EncodingType {
+    BCS,
+    Hex,
+    Base64,
```
Comment (davidiw) [inline config/management/operational/src/keys.rs:59]: so good

File: config/management/operational/src/keys.rs [hunk 5]
```diff
@@ -1,15 +1,221 @@
...
+}
+
+impl FromStr for EncodingType {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "hex" => Ok(EncodingType::Hex),
+            "bcs" => Ok(EncodingType::BCS),
+            "base64" => Ok(EncodingType::Base64),
```
Comment (davidiw) [inline config/management/operational/src/keys.rs:69]: I hate that python has b64 as part of the name of base64 libs... ugh

File: config/management/operational/src/command.rs [hunk 6]
```diff
@@ -26,10 +28,18 @@ pub enum Command {
     CreateValidator(crate::governance::CreateValidator),
     #[structopt(about = "Create a new validator operator account")]
     CreateValidatorOperator(crate::governance::CreateValidatorOperator),
+    #[structopt(about = "Extract peer information from an x25519 PrivateKey file")]
+    ExtractPeerFromFile(crate::keys::ExtractPeerFromFile),
+    #[structopt(about = "Extract peer information from storage")]
+    ExtractPeerFromStorage(crate::keys::ExtractPeerFromStorage),
+    #[structopt(about = "Extract peer information from a list of Public Keys")]
+    ExtractPeersFromKeys(crate::keys::ExtractPeersFromKeys),
```
Comment (davidiw) [inline config/management/operational/src/command.rs:36]: Might be useful to better define what peer information is

File: config/management/operational/src/command.rs [hunk 7]
```diff
@@ -26,10 +28,18 @@ pub enum Command {
     CreateValidator(crate::governance::CreateValidator),
     #[structopt(about = "Create a new validator operator account")]
     CreateValidatorOperator(crate::governance::CreateValidatorOperator),
+    #[structopt(about = "Extract peer information from an x25519 PrivateKey file")]
+    ExtractPeerFromFile(crate::keys::ExtractPeerFromFile),
```
Comment (davidiw) [inline config/management/operational/src/command.rs:32]: might be useful to better define what peer information is -- namely the information required for configuring a trusted network identity.

**Review by davidiw — APPROVED** [review]

**Pull request closed without merging**